### DISCUSSION TITLE
Default locale remapping

### DISCUSSION THREAD

**Discussion by jsebaf** [wintercms/wn-translate-plugin#80] [Q&A] [answered]
Hello.
I'm writing a console command that loads a list of locales and then creates some model translations. The problem is that method setAttributeTranslated() from TranslatableBehavior uses $defaultLocale from Locale that is protected.

I've had to add a method resetDefaultLocale() to Locale that sets $defaultLocale = null and then:

```php
Locale::resetDefaultLocale();
Locale::clearCache();
Translator::forgetInstance();
```
Is there a better solution?
Thank you

**Comment by mjauvin** [comment]
What EXACTLY are you trying to do here?

**Comment by mjauvin** [comment]
You do realize there is a `locale` argument to `setAttributeTranslated()`, right?
```php
public function setAttributeTranslated($key, $value, $locale = null)
```

**Comment by jsebaf** [comment]
I'm trying to load the initial data needed for an application to work. A list of locales (default is non-English) and some translatable data. This executes in one console command.

I'm using setAttributeTranslated() passing the respective locale. But this method uses a value $this->translatableDefault that is obtained from Locale::getDefault(). This static property (defaultLocale) from Locale is protected and hasn't got accessor.

The script then uses old defaultLocale (English) even though the default locale in the database has changed.

**Comment by mjauvin** [comment]
Before inserting the initial data, you should create the different locales using the `Locale` model.
```php
use Winter\Translate\Models\Locale;

$locale = new Locale;
$locale->code = 'fr';
$locale->code = 'Francais';
$locale->is_enabled = 1;
$locale->makeDefault();
$locale->save();
```

Add as many as you need, but only one should be the default, of course.

**Comment by jsebaf** [comment]
The problem is that:
```php
$locale->makeDefault();
```
doesn't change `Locale::$defaultLocale` (used by `setAttributeTranslated()`). This property retains the old default value.

Executing this command on a fresh install:

```php
    public function handle()
    {
        $locale = new Locale;
        $locale->code = 'fr';
        $locale->name = 'Francais';
        $locale->is_enabled = 1;
        $locale->makeDefault();
        $locale->save();

        $this->line(Locale::getDefault()->code);
    }
```
gives output en instead fr.

**Comment by mjauvin** [comment]
Sorry, try saving the new locale first, then calling `makeDefault()` on it.

**Comment by mjauvin** [comment]
No, that's not it either... I had to call `artisan cache:clear` for the new default to kick in... I'll see what needs to be done from within the console command to clear that cache.

**Reply by jsebaf** [reply]
This works if loading locales and loading other data are done in separate commands (clearing cache in the middle). But when done in the same command, `Locale::$defaultLocale` is not updated because it is retrieved from static class storage (not cache).

I think that `makeDefault()` should update `Locale::$defaultLocale`.

**Comment by mjauvin** [comment, marked as answer]
Can you try latest dev-develop version of the plugin, that should help  with your use case.

**Reply by jsebaf** [reply]
I will try it.
Thanks

**Reply by jsebaf** [reply]
The fix works. Thank you!
